polygeist/tools/cgeist/Test/Verification/sycl/functions.cpp

```diff
@@ -678,27 +678,35 @@ SYCL_EXTERNAL void op_1(sycl::id<2> a, sycl::id<2> b) {
-// CHECK-MLIR-NEXT: %4 = "sycl.id.get"(%2, %c1_i32) {BaseType = memref<?x!sycl_array_2_, 4>, FunctionName = @get, MangledFunctionName = @_ZNK4sycl3_V16detail5arrayILi2EE3getEi, TypeName = @array} : (memref<?x!sycl_id_2_, 4>, i32) -> i64
-// CHECK-MLIR-NEXT: %5 = arith.addi %3, %4 : i64
-// CHECK-MLIR-NEXT: %6 = sycl.call(%5) {FunctionName = @abs, MangledFunctionName = @_ZN4sycl3_V13absImEENSt9enable_ifIXsr6detail14is_ugenintegerIT_EE5valueES3_E4typeES3_} : (i64) -> i64
-// CHECK-MLIR-NEXT: return
-// CHECK-MLIR-NEXT: }
+// CHECK-MLIR-NEXT:    %c1_i32 = arith.constant 1 : i32
+// CHECK-MLIR-NEXT:    %c0_i32 = arith.constant 0 : i32
+// CHECK-MLIR-NEXT:    %0 = affine.load %arg0[0] : memref<?x!sycl_id_2_>
+// CHECK-MLIR-NEXT:    %1 = "sycl.id.get"(%0, %c0_i32) {BaseType = memref<?x!sycl_array_2_, 4>, FunctionName = @get, MangledFunctionName = @_ZNK4sycl3_V16detail5arrayILi2EE3getEi, TypeName = @array} : (!sycl_id_2_, i32) -> i64
+// CHECK-MLIR-NEXT:    %2 = "sycl.id.get"(%0, %c1_i32) {BaseType = memref<?x!sycl_array_2_, 4>, FunctionName = @get, MangledFunctionName = @_ZNK4sycl3_V16detail5arrayILi2EE3getEi, TypeName = @array} : (!sycl_id_2_, i32) -> i64
+// CHECK-MLIR-NEXT:    %3 = arith.addi %1, %2 : i64
+// CHECK-MLIR-NEXT:    %4 = sycl.call(%3) {FunctionName = @abs, MangledFunctionName = @_ZN4sycl3_V13absImEENSt9enable_ifIXsr6detail14is_ugenintegerIT_EE5valueES3_E4typeES3_} : (i64) -> i64
+// CHECK-MLIR-NEXT:    return
 
 // CHECK-LLVM-LABEL: define spir_func void @_Z8static_1N4sycl3_V12idILi2EEES2_(
-// CHECK-LLVM            %"class.sycl::_V1::id.2"* noundef byval(%"class.sycl::_V1::id.2") align 8 %0, %"class.sycl::_V1::id.2"* noundef byval(%"class.sycl::_V1::id.2") align 8 %1) #[[FUNCATTRS]]  
-// CHECK-LLVM-NEXT: %3 = addrspacecast %"class.sycl::_V1::id.2"* %0 to %"class.sycl::_V1::id.2" addrspace(4)*
-// CHECK-LLVM-NEXT: %4 = bitcast %"class.sycl::_V1::id.2" addrspace(4)* %3 to %"class.sycl::_V1::detail::array.2" addrspace(4)*
-// CHECK-LLVM-NEXT: %5 = call spir_func i64 @_ZNK4sycl3_V16detail5arrayILi2EE3getEi(%"class.sycl::_V1::detail::array.2" addrspace(4)* %4, i32 0)
-// CHECK-LLVM-NEXT: %6 = call spir_func i64 @_ZNK4sycl3_V16detail5arrayILi2EE3getEi(%"class.sycl::_V1::detail::array.2" addrspace(4)* %4, i32 1)
-// CHECK-LLVM-NEXT: %7 = add i64 %5, %6
-// CHECK-LLVM-NEXT: %8 = call spir_func i64 @_ZN4sycl3_V13absImEENSt9enable_ifIXsr6detail14is_ugenintegerIT_EE5valueES3_E4typeES3_(i64 %7)
-// CHECK-LLVM-NEXT: ret void
+// CHECK-LLVM-SAME:    %"class.sycl::_V1::id.2"* noundef byval(%"class.sycl::_V1::id.2") align 8 %0, %"class.sycl::_V1::id.2"* noundef byval(%"class.sycl::_V1::id.2") align 8 %1) #[[FUNCATTRS]]  
+// CHECK-LLVM-NEXT:   %3 = alloca %"class.sycl::_V1::id.2", align 8
+// CHECK-LLVM-NEXT:   %4 = alloca i64, align 8
```

Comment:
Looks like we can benefit from a mem2reg pass.

Reply:
Yes I am concerned we have 4 alloca, 1 load and 4 stores we did not have before. @victor-eds can you try to add a mem2reg MLIR pass to the pipeline after the lowering?

Reply:
Indeed (or DSE). I checked and, after lowering to LLVM, the problem is fixed there.

Reply:
Sorry, misunderstood. Even after inlining everything, it appears our passes cannot currently do that. Running with `-O1` fixes that, though. I've checked and running a `-inline -sroa` pipeline on the LLVM output fixes this. We have no `-sroa` pass in MLIR, so we cannot currently perform such optimizations. As `-O1` fixes this, I'd say we can keep working on performing optimizations at higher levels and let LLVM do its work on the lower level. Do you agree? @whitneywhtsang @etiotto